### PR TITLE
Expose redirect_uri via postMessage when redirecting

### PR DESCRIPTION
**Pull request by Rubikium**
ref https://github.com/authgear/2023C01-latte-customization/issues/65

### CODE REVIEW

**Comment by louischan-oursky** [comment]
According to https://github.com/authgear/2023C01-latte-customization/issues/65#issuecomment-1450984869 we can just update the template to post message.

**Review by kiootic — COMMENTED** [review]
The custom UI could be at a sub-path of a domain, e.g. `https://example.com/custom-ui/auth`. In that case, developer would set `x_custom_ui_uri: https://example.com/custom-ui/auth`, but the `targetOrigin` of `postMessage` should be `https://example.com`.

Change `CustomUIURI` in the authorization result structs into `CustomUIOrigin` for use of template.

File: pkg/lib/oauth/response_mode.go [hunk 1]
```diff
@@ -21,6 +21,7 @@ const htmlRedirectTemplateString = `<!DOCTYPE html>
 {{- else }}
 <script>
 {{- end }}
+window.parent.postMessage("{{ .redirect_uri }}", "{{ .custom_ui_uri }}")
```
Comment (kiootic) [inline pkg/lib/oauth/response_mode.go:24]: Also, I'd prefer not posting a raw string as a message, but an object for better flexibility. e.g. `{ redirect_uri: "..." }`

File: pkg/util/urlutil/origin.go [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+	"net/url"
+)
+
+func ExtractOrigin(u *url.URL) *url.URL {
+	newU := *u
+	newU.Path = ""
+	newU.RawQuery = ""
+	newU.Fragment = ""
+	return &newU
+}
```
Comment (louischan-oursky) [inline pkg/util/urlutil/origin.go:13]: According to https://pkg.go.dev/net/url#URL and https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin, we only need to copy scheme, opaque and host.

```golang
func ExtractOrigin(u *url.URL) *url.URL {
  return &url.URL{
    Scheme: u.Scheme,
    Opaque: u.Opaque,
    Host: u.Host,
  }
}
```

Write a test for this function.

**Comment by louischan-oursky** [comment]
Try run `make check-tidy` and commit changes.

**Comment by louischan-oursky** [comment]
@Rubikium Have you tested if the message can be delivered to the parent if we have the refresh meta tag with a delay of 0?

**Comment by Rubikium** [comment]
> @Rubikium Have you tested if the message can be delivered to the parent if we have the refresh meta tag with a delay of 0?

Yes, and the message can be delivered with a delay of 0

**Pull request merged**